ui/raidboss/emulator/ui/Tooltip.ts

```diff
@@ -0,0 +1,167 @@
+    if (!arrowElem) {
+      const msg = `Tooltip template direction ${direction} missing .arrow`;
+      console.error(msg);
+      throw new Error(msg);
+    }
+    this.arrow = arrowElem;
+    this.setText(text);
+    document.body.append(this.tooltip);
+    if (autoShow) {
+      showEvents.forEach((e) => {
+        this.target.addEventListener(e, () => {
+          this.show();
+        });
+      });
+    }
+    if (autoHide) {
+      hideEvents.forEach((e) => {
+        this.target.addEventListener(e, () => {
+          this.hide();
+        });
+      });
+    }
+  }
+
+  setText(text: string): void {
+    this.inner.textContent = text;
+  }
+
+  show(): void {
+    const targetRect = this.target.getBoundingClientRect();
+    const targetMiddle = {
+      x: targetRect.x + (targetRect.width / 2),
+      y: targetRect.y + (targetRect.height / 2),
+    };
+    const tooltipRect = this.tooltip.getBoundingClientRect();
+    // Middle of tooltip - half of arrow height
+    const lrArrowHeight = (tooltipRect.height / 2) -
+      (this.arrow.getBoundingClientRect().height / 2);
+    switch (this.direction) {
+    case 'top':
+      this.tooltip.style.left = toPx((targetMiddle.x - (tooltipRect.width / 2)) + this.offset.x);
+      this.tooltip.style.bottom = toPx((targetRect.y - tooltipRect.height) + this.offset.y);
+      break;
+    case 'right':
+      this.tooltip.style.left = toPx(targetRect.right + this.offset.x);
+      this.tooltip.style.top = toPx((targetMiddle.y - (tooltipRect.height / 2)) + this.offset.y);
+      this.arrow.style.top = toPx(lrArrowHeight);
+      break;
+    case 'bottom':
+      this.tooltip.style.left = toPx((targetMiddle.x - (tooltipRect.width / 2)) + this.offset.x);
+      this.tooltip.style.top = toPx(targetRect.bottom + this.offset.y);
+      break;
+    case 'left':
+      this.tooltip.style.left = toPx((targetRect.left - tooltipRect.width) + this.offset.x);
+      this.tooltip.style.top = toPx((targetMiddle.y - (tooltipRect.height / 2)) + this.offset.y);
+      this.arrow.style.top = toPx(lrArrowHeight);
+      break;
+    }
+    this.tooltip.classList.add('show');
+    this.tooltip.setAttribute('data-show', '');
+  }
+
+  hide(): void {
+    this.tooltip.classList.remove('show');
+    this.tooltip.removeAttribute('data-show');
+  }
+
+  static initializeTemplates(): void {
+    if (Tooltip.templates)
+      return;
+
+    Tooltip.templates = {
+      top: Tooltip.getTemplate('top'),
+      right: Tooltip.getTemplate('right'),
+      bottom: Tooltip.getTemplate('bottom'),
+      left: Tooltip.getTemplate('left'),
+    };
+  }
+
+  static getTemplate(dir: string): HTMLTemplateElement {
+    const elemName = `${dir}TooltipTemplate`;
+    const ret = document.getElementById(elemName);
+    if (ret === null) {
+      const msg = `Missing element ${elemName}`;
+      console.error(msg);
+      throw new Error(msg);
+    }
+    return ret as HTMLTemplateElement;
```

Comment:
```suggestion
    if (ret && ret instanceof HTMLTemplateElement)
      return ret;

    const msg = `Missing element ${elemName}`;
    console.error(msg);
    throw new Error(msg);
```

Reply:
I don't think `instanceof` will work here, since the abstract declaration is just a way of tricking tsc into letting this work?

Or it could be checking `instanceof HTMLElement` instead?

![image](https://user-images.githubusercontent.com/6119598/120078884-846c4600-c07f-11eb-936f-16aedbeb14c3.png)

Reply:
Oh fair.  `instance of HTMLElement` seems better then, if that's what it really is.

Reply:
It's `instanceof HTMLTemplateElement` but there's no way to get tsc to see the browser-generated-at-runtime type, as far as I'm aware.